pkg/util/testutil/patternscanner.go

```diff
@@ -61,17 +76,41 @@ func (ps *PatternScanner) Write(p []byte) (n int, err error) {
 	// Process all complete lines.
 	for _, line := range lines[:len(lines)-1] {
 		ps.buffers = append(ps.buffers, line) // Save the log line.
-		if !ps.stopped && ps.pattern.MatchString(line) {
-			ps.stopOnce.Do(func() {
-				ps.stopped = true
-				close(ps.DoneChan) // Notify the caller about the match.
-			})
-		}
+
+		// Check if we've met the scanning criteria
+		ps.matchPatterns(line)
 	}
 
 	return len(p), nil
 }
 
+// matchPatterns checks if the current line matches the scanning requirements
+func (ps *PatternScanner) matchPatterns(line string) {
+	switch {
+	// startPatternFound pattern not found yet, look for it
+	case !ps.startPatternFound:
+		if ps.startPattern.MatchString(line) {
+			// found start pattern, flip the flag to start looking for finish pattern for following iterations
+			ps.startPatternFound = true
+
+			// no finishPattern provided, we can stop here
+			if ps.finishPattern == nil {
+				ps.notifyAndStop()
+			}
+		}
+	// startPatternFound pattern found, look for finish pattern if provided
+	case ps.finishPattern != nil && ps.finishPattern.MatchString(line):
+		ps.notifyAndStop()
```

Comment:
I am not sure I understand how is this going to be used. IIUC, this will only notify when the start pattern is found and then the finish pattern is found, right? If the log messages are in order this is equivalent to waiting for just the finish pattern, no?

Reply:
it will notify when all provided (at least one is required) patterns are found.
if both provided, then both must be matched before the DoneChan is triggered. 
if we wait just for the finishPattern, we might not get the startPattern at all (which is not the expected behavior)